klippy/extras/adxl345.py

```diff
@@ -170,23 +193,35 @@ def _convert_sequence(self, sequence):
         if sequence < self.last_sequence:
             sequence += 0x10000
         return sequence
+    def read_reg(self, reg):
+        params = self.spi.spi_transfer([reg | REG_MOD_READ, 0x00])
+        response = bytearray(params['response'])
+        return response[1]
+    def set_reg(self, reg, val, minclock=0):
+        self.spi.spi_send([reg, val & 0xFF], minclock=minclock)
+        stored_val = self.read_reg(reg)
+        if stored_val != val:
+            raise self.printer.command_error(
+                    "Failed to set ADXL345 register [0x%x] to 0x%x: got 0x%x. "
+                    "This is generally indicative of connection problems "
+                    "(e.g. faulty wiring) or a faulty adxl345 chip." % (
+                        reg, val, stored_val))
+    def is_measuring(self):
+        return self.query_rate > 0
     def start_measurements(self, rate=None):
+        if self.is_measuring():
+            return
         rate = rate or self.data_rate
-        # Verify chip connectivity
-        params = self.spi.spi_transfer([REG_DEVID | REG_MOD_READ, 0x00])
-        response = bytearray(params['response'])
-        if response[1] != 0xe5:
-            raise self.printer.command_error("Invalid adxl345 id (got %x vs %x)"
-                                             % (response[1], 0xe5))
+        if not self.is_initialized():
+            self.initialize()
```

Comment:
If the spi connection isn't wired correctly and all results are returning 0xff, would this pass without error?  That is, would `is_initialized()` return True because the read to REG_DATA_FORMAT returns non-zero, and thus skip the 0xe5 check in `initalize()`.

Reply:
That's fair. I added the check for 0xe5 device id to is_initialized too. I kept raising the exception in a single place though.